services/blockchain-indexer/shared/dataService/business/pos/stakers.js

```diff
@@ -0,0 +1,115 @@
+/*
+ * LiskHQ/lisk-service
+ * Copyright © 2022 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const BluebirdPromise = require('bluebird');
+
+const {
+	MySQL: { getTableInstance },
+} = require('lisk-service-framework');
+
+const config = require('../../../../config');
+const stakesIndexSchema = require('../../../database/schema/stakes');
+
+const {
+	updateAccountPublicKey,
+	getIndexedAccountInfo,
+	getLisk32AddressFromPublicKey,
+} = require('../../../utils/accountUtils');
+
+const MYSQL_ENDPOINT = config.endpoints.mysql;
+
+const getStakesIndex = () => getTableInstance(
+	stakesIndexSchema.tableName,
+	stakesIndexSchema,
+	MYSQL_ENDPOINT,
+);
+
+const getPoSStakers = async params => {
+	const stakesTable = await getStakesIndex();
+	const stakers = {
+		data: { stakers: [] },
+		meta: {},
+	};
+
+	if (params.address) {
+		const { address, ...remParams } = params;
+		params = remParams;
+
+		params.validatorAddress = address;
+	}
+
+	if (params.publicKey) {
+		const { publicKey, ...remParams } = params;
+		params = remParams;
+
+		params.validatorAddress = getLisk32AddressFromPublicKey(publicKey);
+
+		// Index publicKey
+		await updateAccountPublicKey(publicKey);
+	}
+
+	if (params.name) {
+		const { name, ...remParams } = params;
+		params = remParams;
+
+		const { address } = await getIndexedAccountInfo({ name, limit: 1 }, ['address']);
+		if (!address) return new Error(`Account with name: ${name} does not exist.`);
+		params.validatorAddress = address;
+	}
+
+	// TODO: Use count method directly once support for custom column-based count added https://github.com/LiskHQ/lisk-service/issues/1188
+	const [{ numStakers }] = await stakesTable.rawQuery(`SELECT COUNT(*) as numStakers from ${stakesIndexSchema.tableName} WHERE receivedAddress='${params.receivedAddress}'`);
+
+	const resultSet = await stakesTable.find(
+		{
+			validatorAddress: params.validatorAddress,
+			limit: numStakers,
+		},
+		Object.keys(stakesIndexSchema.schema),
+	);
+	if (resultSet.length) stakers.data.stakers = resultSet;
+
+	stakers.data.stakers = await BluebirdPromise.map(
+		stakers.data.stakers,
+		async stake => {
+			const { name = null } = await getIndexedAccountInfo({ address: stake.stakerAddress, limit: 1 }, ['name']);
```

Comment:
Incase name is not found in the table, should we also look into the delegate cache?

Reply:
No. All the names should be available in the table. If it's no present there, we need to fix that in a separate issue.